### PR TITLE
Add pytest-based test discovery and runnables for Python

### LINKED ISSUE #12080: Add test discovery and runnables for Python files
### Check for existing issues

- [X] Completed

### Describe the feature

There are a load of python testing frameworks, but of those, the most popular are:

- [X] [`unittest`](https://docs.python.org/3/library/unittest.html) - Python's stdlib testing framework
    - [Test discovery rules](https://docs.python.org/3/library/unittest.html#unittest-test-discovery)
- [ ] [`pytest`](https://docs.pytest.org/en/8.2.x/) - 3rd party, but likely the most-used testing framework
    - [Test discovery rules](https://docs.pytest.org/en/8.2.x/explanation/goodpractices.html#conventions-for-python-test-discovery)

It would be great if we could offer testing support for both `unittest` and `pytest`, as I think that would cover most users. We should include commands to run all tests, for both frameworks, and to discover and run single tests, for both frameworks. All that being said, I'm not sure if our task infrastructure is really set up to handle being able to run tests from different testing frameworks. @osiewicz would know more.

---

Honorable mentions for Python testing frameworks go to [ward](https://github.com/darrenburns/ward) - its modern, neat, and then one I reach for these days.

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by jrderuiter**
Closes  #12080, #18649.

Screenshot:

image

Still in progress:

1. I'd like to add configuration options for selecting a Python test runner (either pytest or unittest) so that users can explicitly choose which runner they'd like to use for running their tests. This preference has to be configured as unittest-style tests can also be run by pytest, meaning we can't rely on auto-discovery to choose the desired test runner.
2. I'd like to add venv auto-discovery similar to the feature currently provided by the terminal using detect_venv. 
3. Unit tests.

Unfortunately I'm struggling a bit with how to add settings in the appropriate location (e.g. Python language settings). Can anyone provide me with some pointers and/or examples on how to either add extra settings or to re-use the existing ones?

My rust programming level is OK-ish but I'm not very familiar with the Zed project structure and could use some help.

I'm also open for pair programming as mentioned on the website if that helps!

Release Notes:

- Added pytest-based test discovery and runnables for Python.
- Adds a configurable option for switching between unittest and pytest as a test runner under Python language settings. Set "TASK_RUNNER" to "unittest" under task settings for Python if you wish to use unittest to run Python tasks; the default is pytest.

### CODE REVIEW

**Comment by jrderuiter** [comment]
Still in progress:

1. I'd like to add configuration options for selecting a Python test runner (either pytest or unittest) so that users can explicitly choose which runner they'd like to use for running their tests. This preference has to be configured as unittest-style tests can also be run by pytest, meaning we can't rely on auto-discovery to choose the desired test runner.
2. I'd like to add venv auto-discovery similar to the feature currently provided by the terminal using detect_venv. 

Unfortunately I'm struggling a bit with how to add settings in the appropriate location (e.g. Python configuration options). Can anyone provide me with some pointers and/or examples on how to either add extra settings or to re-use the existing ones?

My rust programming level is OK-ish but I'm not very familiar with the Zed project structure and could use some help.

I'm also open for pair programming as mentioned on the website if that helps :)

**Comment by NeilRiver** [comment]
oh my god finally if this happens i will go to kostel and never laugh at beavers in poland again

**Comment by jrderuiter** [comment]
@osiewicz Any feedback/pointers?

Or should I move my questions to the community channel?

File: crates/languages/src/python/runnables.scm [hunk 1]
```diff
@@ -29,3 +29,42 @@
             )
         )
 )
+
+; pytest functions
+(
+    (module
+        (function_definition
+            name: (identifier) @run @_pytest_method_name
+            (#match? @_pytest_method_name "^test_.*")
```
Comment (osiewicz) [inline crates/languages/src/python/runnables.scm:38]: This regex can be a bit more lenient:
```suggestion
            (#match? @_pytest_method_name "^test_")
```

File: crates/languages/src/python/runnables.scm [hunk 2]
```diff
@@ -29,3 +29,42 @@
...
+    )
+    (#set! tag python-pytest-method)
+)
+
+; pytest classes
+(
+    (module
+        (class_definition
+            name: (identifier) @run @_pytest_class_name
+            (#match? @_pytest_class_name "^Test.*")
```
Comment (osiewicz) [inline crates/languages/src/python/runnables.scm:49]: Same goes for this one:
```suggestion
            (#match? @_pytest_class_name "^Test")
```

File: crates/languages/src/python/runnables.scm [hunk 3]
```diff
@@ -29,3 +29,42 @@
...
+        (#set! tag python-pytest-class)
+    )
+)
+
+; pytest class methods
+(
+    (module
+        (class_definition
+            name: (identifier) @_pytest_class_name
+            (#match? @_pytest_class_name "^Test.*")
```
Comment (osiewicz) [inline crates/languages/src/python/runnables.scm:60]:
```suggestion
            (#match? @_pytest_class_name "^Test")
```

File: crates/languages/src/python/runnables.scm [hunk 4]
```diff
@@ -29,3 +29,42 @@
...
+            body: (block
+                    (function_definition
+                        name: (identifier) @run @_pytest_method_name
+                        (#match? @_pytest_method_name "^test.*")
```
Comment (osiewicz) [inline crates/languages/src/python/runnables.scm:64]:
```suggestion
                        (#match? @_pytest_method_name "^test")
```

File: crates/languages/src/python.rs [hunk 5]
```diff
@@ -240,11 +245,120 @@ impl ContextProvider for PythonContextProvider {
...
+
+        Ok(task::TaskVariables::from_iter([test_target]))
+    }
+
+    fn associated_tasks(
+        &self,
+        _: Option<Arc<dyn language::File>>,
+        _cx: &AppContext,
+    ) -> Option<TaskTemplates> {
+        let python_path = ".venv/bin/python"; // TODO: Re-use terminal detect_venv logic for this. (How?)
```
Comment (osiewicz) [inline crates/languages/src/python.rs:263]: To touch on this: I'll work on making this work with virtual environments. Let's not worry about it in this PR.

File: crates/languages/src/python.rs [hunk 6]
```diff
@@ -240,11 +245,120 @@ impl ContextProvider for PythonContextProvider {
...
+        let test_runner = TestRunner::PYTEST; // TODO: Get this from settings. (How?)
```
Comment (osiewicz) [inline crates/languages/src/python.rs:264]: You can see how Rust does it:
https://github.com/zed-industries/zed/blob/695176898e809fe097b5f6ea6db5712cf1cfefe0/crates/languages/src/rust.rs#L427

File: crates/languages/src/python.rs [hunk 7]
```diff
@@ -255,60 +369,47 @@ impl ContextProvider for PythonContextProvider {
             }
             (Some(class_name), None) => format!("{}.{}", python_module_name, class_name),
             (None, None) => python_module_name,
-            (None, Some(_)) => return Ok(task::TaskVariables::default()), // should never happen, a TestCase class is the unit of testing
+            (None, Some(_)) => return Ok((VariableName::Custom(Cow::Borrowed("")), String::new())), // should never happen, a TestCase class is the unit of testing
```
Comment (osiewicz) [inline crates/languages/src/python.rs:372]: Since this function returns Result, we could probably get away with returning an Error instead?

**Comment by osiewicz** [comment]
@jrderuiter  Hey, sorry for slacking. I left a couple of comments, but overall your PR looks good :)

**Comment by NeilRiver** [comment]
@osiewicz pane dziekuje
@jrderuiter well done! keep it up!

**Comment by NeilRiver** [comment]
@osiewicz maybe you can continue? I have a suspicion that this ticket will never be completed

**Comment by jrderuiter** [comment]
I’ve been away but will pick this up again in a week time.

**Comment by NeilRiver** [comment]
> I’ve been busy but will pick this up again in a week time.

@jrderuiter cool! thanks

**Comment by NeilRiver** [comment]
I ask you to look at the bug from vscode, when you launch a test via the launch icon (opposite the function) and debug it, then click on restart from above, instead of restarting this test, God knows what restarts, and you will need to interrupt the test and click on the start icon again.

**Comment by jrderuiter** [comment]
@osiewicz I was looking at the work from your PR (https://github.com/zed-industries/zed/pull/19576) but I'm still struggling to see how I can get the selected toolchain Python path inside the `associated_tasks` function. Do you maybe have an example of this?

**Comment by osiewicz** [comment]
@jrderuiter Sorry, I didn't land a PR for that yet. We could follow up on this PR once that lands.

**Comment by jrderuiter** [comment]
Pyright already uses the toolchain selector though right? How is that passed along exactly?

**Comment by osiewicz** [comment]
That's a bit different, as PyRight is a language server and toolchains are already wired through there.

**Comment by osiewicz** [comment]
Ok, https://github.com/zed-industries/zed/pull/20667/files has landed, so you can use `PYTHON_ACTIVE_TOOLCHAIN_PATH.template_value()` as your executable and that ought to just work.

File: crates/languages/src/python.rs [hunk 8]
```diff
@@ -259,8 +264,8 @@ async fn get_cached_server_binary(
 
 pub(crate) struct PythonContextProvider;
 
-const PYTHON_UNITTEST_TARGET_TASK_VARIABLE: VariableName =
-    VariableName::Custom(Cow::Borrowed("PYTHON_UNITTEST_TARGET"));
+const PYTHON_TEST_TARGET_TASK_VARIABLE: VariableName =
```
Comment (osiewicz) [inline crates/languages/src/python.rs:267]: On one hand this is a breaking change, but on the other I don't think we have this variable documented anywhere. We should be good.

**Comment by jrderuiter** [comment]
Sure, go ahead! Do you need access to my branch?

**Comment by osiewicz** [comment]
Nope, I'm good :) I think maintainers of a repo get write access to forks by default.

**Comment by jrderuiter** [comment]
Looks good, thanks for the help in finishing the PR!

**Comment by Tankanow** [comment]
@jrderuiter and @osiewicz, this is amazing. Thank you very much for your contributions. I can't wait until this is released!

**Comment by NeilRiver** [comment]
@osiewicz powiedz mi, czy debugowanie działa? czy to tylko uruchomienie i nic więcej? dziękuję

**Comment by osiewicz** [comment]
@NeilRiver could you please use English like everybody else? You don't need to use Polish when pinging me.
No, it does not work with debugging yet. That's a separate issue.

**Comment by absynce** [comment]
There's a typo in the release notes for the configuration variable, "TASK_RUNNER" should be "TEST_RUNNER".

To save folks time, here's my relevant setting, which reverts to the previous default runner, `unittest`:
```json
{
  "Python": {
    "tasks": {
      "variables": {
        "TEST_RUNNER": "unittest"
      }
    }
  }
}
```

Big thanks to y'all for continuing to make Zed awesome! ❤️

**Pull request merged**